### PR TITLE
Add support for generating ResourceId and InputObject parameter sets for Azure specs

### PR DESCRIPTION
**Pull request by bmanikm**
Added support for generating ResourceId and InputObject parameter sets, along with 'Name' as the ResourceName parameter name for AzureRM swagger specs.

For AzureRM swagger specs, when UseAzureCsharpGenerator switch is specified and the resource id follows the guidelines of Azure Resource operations:
- The following additional parameter sets will be generated
    - InputObject parameter set with the same object type returned by Get. Supports piping from Get operarion to action cmdlets.
    - ResourceId parameter set which splits the resource id into component parts (supports piping from generic cmdlets).
- Parameter name of Azure resource name parameter will be generated as 'Name' and the actual resource name parameter from the resource id will be added as an alias.

The following pipeline operations are manually validated for the Azure CosmosDB RP.
```powershell
# -Name parameter for ResourceName
$DatabaseAccount = Get-DatabaseAccount -Name $ExistingAccountName -ResourceGroupName $ResourceGroupName -Verbose
# -AccountName parameter which is an alias to Name parameter
$DatabaseAccount = Get-DatabaseAccount -AccountName $ExistingAccountName -ResourceGroupName $ResourceGroupName

# ResourceId parameter set
Get-DatabaseAccount -ResourceId $DatabaseAccount.Id
# ResourceId parameter set by piping the output generic Get-AzureRmResource cmdlet
Get-AzureRmResource -ResourceId $DatabaseAccount.Id | Get-DatabaseAccount

# InputObject parameter set
Get-DatabaseAccount -InputObject $DatabaseAccount
# InputObject parameter set by piping the DatabaseAccount object
$DatabaseAccount | Get-DatabaseAccount
```

### CODE REVIEW

File: PSSwagger/Get-ArmResourceIdParameterValue.ps1 [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    
+    if ($IdTokens.Count -ne $IdTemplateTokens.Count) {
+        Write-Error -Message ("Invalid resource id '{0}'. Expected id template is '{1}'." -f $Id, $IdTemplate) -ErrorId 'InvalidResourceIdValue'
+        return
+    }
+    
+    $ResourceIdParameterValues = [ordered]@{}
+    for ($i = 0; $i -lt $IdTemplateTokens.Count; $i++) {
+        if ($IdTokens[$i] -ne $IdTemplateTokens[$i]) {
+            if ($IdTemplateTokens[$i] -notmatch '{*}') {
```
Comment (brywang-msft) [inline PSSwagger/Get-ArmResourceIdParameterValue.ps1]: >{*} [](start = 49, length = 3)

>{*} [](start = 49, length = 3)

>{*} [](start = 49, length = 3)

What's the goal of this pattern in plain English? Assuming it's what I think, should be: '^\{.*\}$'.

'{blah}' -replace '{*}','x' should result in '{blahx' #Closed

File: PSSwagger/Get-ArmResourceIdParameterValue.ps1 [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+.PARAMETER  IdTemplate
+    The unique resource identifier template of the Azure Resource.
+#>
+function Get-ArmResourceIdParameterValue {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [string]
+        $Id,
```
Comment (brywang-msft) [inline PSSwagger/Get-ArmResourceIdParameterValue.ps1:18]: >$Id [](start = 8, length = 3)

Write ID to verbose/debug? #Closed

File: PSSwagger/Paths.psm1 [hunk 3]
```diff
@@ -208,6 +220,13 @@ function Get-SwaggerSpecPathInfo
                     if($_.Value.ContainsKey('Mandatory') -and $_.Value.Mandatory -eq '$true') {
                         $Priority++
                     }
+
+                    # Add alias for the resource name parameter.
+                    if($ResourceIdAndInputObjectDetails -and
```
Comment (brywang-msft) [inline PSSwagger/Paths.psm1:233]: >if [](start = 20, length = 2)

Also need to ensure that no other parameters are named 'Name' #Closed

File: PSSwagger/SwaggerUtils.psm1 [hunk 4]
```diff
@@ -1705,14 +1756,104 @@ function Get-OutputType
...
+        Write-Debug "Get operation is available in $ResourceId"
+        return
+    }
+
+    $tokens = $ResourceId.Split('/', [System.StringSplitOptions]::RemoveEmptyEntries)
+    if (-not $tokens -or ($tokens.Count -lt 8)) {
+        Write-Debug -Message ("The specified endpoint '{0}' is not a valid resource identifier." -f $ResourceId)
+        return
+    }
+    $ResourceIdParameters = $tokens | Where-Object {$_ -match '{*}'} | Foreach-Object {
```
Comment (brywang-msft) [inline PSSwagger/SwaggerUtils.psm1]: >{*} [](start = 63, length = 3)

>{*} [](start = 63, length = 3)

>{*} [](start = 63, length = 3)

Same regex comments as in Get-ArmResourceIdParameterValue.ps1 #Closed

File: Tests/PSSwagger.Unit.Tests.ps1 [hunk 5]
```diff
@@ -118,21 +118,24 @@ Describe "PSSwagger Unit Tests" -Tag @('BVT', 'DRT', 'UnitTest', 'P0') {
...
                     IsWindows = $false
                 }
             }
 
             Mock Get-EnvironmentVariable -ModuleName PSSwaggerUtility {
                 if ('HOME' -eq $Name) {
                     return '/hometest'
-                } elseif ('XDG_CACHE_HOME' -eq $Name) {
+                }
+                elseif ('XDG_CACHE_HOME' -eq $Name) {
```
Comment (brywang-msft) [inline Tests/PSSwagger.Unit.Tests.ps1:132]: What did you use to format? #Closed

**Review by brywang-msft — CHANGES_REQUESTED** [review]
See previous for requested changes

File: Tests/PSSwagger.Unit.Tests.ps1 (hunk 5 above)
Comment (bmanikm) [reply Tests/PSSwagger.Unit.Tests.ps1:132]: Formatting option in VS Code.

---
In reply to: [147846907](https://github.com/PowerShell/PSSwagger/pull/370#discussion_r147846907) [](ancestors = 147846907)

**Review by brywang-msft — APPROVED** [review]
